File: 11-container-with-most-water/11-container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
      int l = 0, h=height.size()-1;
      int res = 0;
      while(l < h){
        if((h-l)*min(height[h],height[l]) > res)
          res = (h-l)*min(height[h],height[l]);
        if(height[l] < height[h]){
          int s = l+1;
          while(s < h && height[s] <= height[l])s++;
          l=s;
        } else {
          int s = h-1;
          while(s > l && height[s] <= height[h])s--;
          h=s;
        }  
      }
      return res;
    }
};

File: 11-container-with-most-water/11-container-with-most-water.scala
object Solution {
    def maxArea(height: Array[Int]): Int = {
      val sz = height.size
      if(sz == 0) return 0

      var lmax = 0
      var rmax = sz - 1
      var max = 0
      while(lmax < rmax){
        if(height(lmax) > height(rmax)){
          val water = height(rmax)*(rmax-lmax)
          if(water > max)max = water
          rmax = rmax-1
        } else {
          val water = height(lmax)*(rmax-lmax)
          if(water > max)max = water
          lmax = lmax+1
        }
        
      }
      max
    }
}